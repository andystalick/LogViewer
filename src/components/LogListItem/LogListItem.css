.log-list-item {
  font-size: 0.8125rem;
  font-family: var(--fontFamilyCode);
  display: flex;
  flex-direction: column;
  white-space: nowrap;
  overflow-y: auto;

  &:nth-of-type(2n) {
    background: var(--list-zebra);
  }

  &[open] {
    background-color: var(--selected);
  }

  summary {
    padding: 0.5rem 0.5rem 0.5rem 0.5rem;
    cursor: pointer;
    display: flex;
    flex-direction: row;
    gap: 1rem;
    &::before {
      content: '▸';
      font-size: 1.2rem;
      line-height: 0.6;
      color: var(--neutral50);
    }
    .item-raw {
      overflow-x: hidden;
    }
  }
  &[open] summary {
    &::before {
      content: '▾';
    }
  }
}
