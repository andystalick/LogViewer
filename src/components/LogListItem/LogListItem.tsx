// import './LogListItem.css';

import { memo, useState } from 'react';

import JsonDisplay from '../JsonDisplay/JsonDisplay';

interface LogListItemProps {
  rawRow: string;
  rawTime: string;
  open?: boolean;
}

const LogListItem = memo<LogListItemProps>((props) => {
  const { rawRow, rawTime, open } = props;
  const formattedTime = new Date(Number(rawTime)).toISOString();

  const [isOpen, setIsOpen] = useState(open == undefined ? false : open);

  const handleClick = () => {
    setIsOpen(!isOpen);
  };

  return (
    <div
      className={`log-list-item ${isOpen && 'open'}`}
      data-testid="log-list-item"
    >
      <div className="item-summary" onClick={handleClick} role="button">
        <span>{formattedTime}</span>
        <span className="item-raw">{rawRow}</span>
      </div>
      {isOpen && <JsonDisplay str={rawRow} />}
    </div>
  );
});

export default LogListItem;
