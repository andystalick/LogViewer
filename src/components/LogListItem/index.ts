import LogListItem from './LogListItem';

export default LogListItem;
