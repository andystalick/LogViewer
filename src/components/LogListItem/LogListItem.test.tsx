import { expect, test } from 'vitest';
import { render, screen } from '@testing-library/react';

import LogListItem from './LogListItem';

const rawTimeFixture = '1725038130163';
const rawRowFixture =
  '{"_time":1722523685074,"cid":"api","channel":"conf:policies","level":"info","message":"loading policy","context":"cribl","policy":{"args":["groupName","dashboardId"],"template":["GET /m/${groupName}/search/dashboards/${dashboardId}","GET /m/${groupName}/search/dashboards/${dashboardId}/*"],"description":"Can view and use the dashboard","title":"Read Only"}}';

test('snapshot test closed', async () => {
  render(<LogListItem rawTime={rawTimeFixture} rawRow={rawRowFixture} />);
  expect(screen.getByTestId('log-list-item')).toMatchSnapshot();
});

//we trust that the browser can toggle Details open and closed and handle clicks on Summary, but we want snapshots to prove the child is rendering conditionally
test('snapshot test open', async () => {
  render(
    <LogListItem rawTime={rawTimeFixture} rawRow={rawRowFixture} open={true} />
  );
  expect(screen.getByTestId('json-display')).toBeDefined();
});
