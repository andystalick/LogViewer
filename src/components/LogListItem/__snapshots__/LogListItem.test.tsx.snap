// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`snapshot test closed 1`] = `
<div
  class="log-list-item false"
  data-testid="log-list-item"
>
  <div
    class="item-summary"
    role="button"
  >
    <span>
      2024-08-30T17:15:30.163Z
    </span>
    <span
      class="item-raw"
    >
      {"_time":1722523685074,"cid":"api","channel":"conf:policies","level":"info","message":"loading policy","context":"cribl","policy":{"args":["groupName","dashboardId"],"template":["GET /m/\${groupName}/search/dashboards/\${dashboardId}","GET /m/\${groupName}/search/dashboards/\${dashboardId}/*"],"description":"Can view and use the dashboard","title":"Read Only"}}
    </span>
  </div>
</div>
`;
